secret-sync-controller/controllers/secretsync_controller.go

```diff
@@ -0,0 +1,496 @@
+	}
+
+	secretObj := ss.Spec.SecretObject
+	if err = secretutil.ValidateSecretObject(secretName, secretObj); err != nil {
+		logger.Error(err, "failed to validate secret object", "secretName", secretName)
+		r.updateStatusConditions(ctx, ss, ConditionTypeUnknown, conditionType, ConditionReasonUserInputValidationFailed, true)
+		return ctrl.Result{}, err
+	}
+
+	secretType := secretutil.GetSecretType(strings.TrimSpace(secretObj.Type))
+	var datamap map[string][]byte
+	if datamap, err = secretutil.GetSecretData(secretObj.Data, secretType, files); err != nil {
+		logger.Error(err, "failed to get secret data", "secretName", secretName)
+		r.updateStatusConditions(ctx, ss, ConditionTypeUnknown, conditionType, ConditionReasonUserInputValidationFailed, true)
+		return ctrl.Result{}, err
+	}
+
+	labels := make(map[string]string)
+	if secretObj.Labels != nil {
+		labels = secretObj.Labels
+	}
+	if val, ok := labels[ControllerLabelKey]; ok {
+		// if the label is already present, check if it is the same as the controller label value
+		// if it is not the same, return an error
+		if val != ControllerLabelValue {
+			r.updateStatusConditions(ctx, ss, ConditionTypeUnknown, conditionType, ConditionReasonFailedInvalidLabelError, true)
+			return ctrl.Result{}, fmt.Errorf("label %s is reserved for use by the secret sync controller", ControllerLabelKey)
+		}
+	} else {
+		// if the label is not present, add it
+		labels[ControllerLabelKey] = ControllerLabelValue
+	}
+
+	annotations := make(map[string]string)
+	if secretObj.Annotations != nil {
+		annotations = secretObj.Annotations
+	}
+	if val, ok := annotations[ControllerAnnotationKey]; ok {
+		// if the annotation is already present, check if it is the same as the controller annotation value
+		// if it is not the same, return an error
+		if val != ControllerLabelValue {
+			r.updateStatusConditions(ctx, ss, ConditionTypeUnknown, conditionType, ConditionReasonFailedInvalidAnnotationError, true)
+			return ctrl.Result{}, fmt.Errorf("annotation %s is reserved for use by the secret sync controller", ControllerAnnotationValue)
+		}
+	} else {
+		// if the annotation is not present, add it
+		annotations[ControllerAnnotationKey] = ControllerAnnotationValue
+	}
+
+	// Compute the hash of the secret
+	syncHash, err := r.computeSecretDataObjectHash(datamap, spc, ss)
+	if err != nil {
+		logger.Error(err, "failed to compute secret data object hash", "secretName", secretName)
+		return ctrl.Result{}, err
+	}
+
+	// Check if the hash has changed.
+	hashChanged := false
+	if syncHash != ss.Status.SyncHash {
+		hashChanged = true
+	}
+
+	// Check if a secret create or update failed and if the controller should re-try the operation
+	failedCondition := metav1.Condition{}
+	for _, ssCondition := range ss.Status.Conditions {
+		if slices.Contains(FailedConditionsTriggeringRetry, ssCondition.Reason) {
+			failedCondition = ssCondition
+			break
+		}
+	}
+
+	if failedCondition.Type != "" || hashChanged {
```

Comment:
If `len(failedCondition.Type) == 0 && !hashChanged` we can return early. Then we don't need this large chunk of code inside the if block.